cpp/src/arrow/dataset/file_base.h

```diff
@@ -45,82 +46,150 @@ namespace dataset {
-  SourceKind id() const { return static_cast<SourceKind>(impl_.index()); }
+  FileSource() : impl_(CustomOpen{&InvalidOpen}) {}
 
   /// \brief Return the type of raw compression on the file, if any
   Compression::type compression() const { return compression_; }
 
-  /// \brief Whether the this source may be opened writable
-  bool writable() const { return writable_; }
-
-  /// \brief Return the file path, if any. Only valid when file source
-  /// type is PATH
+  /// \brief Return the file path, if any. Only valid when file source wraps a path.
   const std::string& path() const {
-    static std::string buffer_path = "<Buffer>";
-    return id() == PATH ? util::get<PATH>(impl_).path : buffer_path;
+    if (IsPath()) {
+      return util::get<PathAndFileSystem>(impl_).path;
+    }
+    if (IsBuffer()) {
+      static std::string no_path = "<Buffer>";
+      return no_path;
+    } else {
+      static std::string no_path = "<CustomOpen>";
+      return no_path;
+    }
   }
 
-  /// \brief Return the filesystem, if any. Only non null when file
-  /// source type is PATH
+  /// \brief Return the filesystem, if any. Otherwise returns nullptr
   const std::shared_ptr<fs::FileSystem>& filesystem() const {
-    static std::shared_ptr<fs::FileSystem> no_fs = NULLPTR;
-    return id() == PATH ? util::get<PATH>(impl_).filesystem : no_fs;
+    if (!IsPath()) {
+      static std::shared_ptr<fs::FileSystem> no_fs = NULLPTR;
+      return no_fs;
+    }
+    return util::get<PathAndFileSystem>(impl_).filesystem;
   }
 
   /// \brief Return the buffer containing the file, if any. Only value
   /// when file source type is BUFFER
   const std::shared_ptr<Buffer>& buffer() const {
-    static std::shared_ptr<Buffer> path_buffer = NULLPTR;
-    return id() == BUFFER ? util::get<BUFFER>(impl_) : path_buffer;
+    if (!IsBuffer()) {
+      static std::shared_ptr<Buffer> no_buffer = NULLPTR;
+      return no_buffer;
+    }
+    return util::get<std::shared_ptr<Buffer>>(impl_);
   }
 
   /// \brief Get a RandomAccessFile which views this file source
-  Result<std::shared_ptr<arrow::io::RandomAccessFile>> Open() const;
+  Result<std::shared_ptr<io::RandomAccessFile>> Open() const;
+
+ private:
+  static Result<std::shared_ptr<io::RandomAccessFile>> InvalidOpen() {
+    return Status::Invalid("Called Open() on an uninitialized FileSource");
+  }
+
+  bool IsPath() const { return util::holds_alternative<PathAndFileSystem>(impl_); }
+
+  bool IsBuffer() const {
+    return util::holds_alternative<std::shared_ptr<Buffer>>(impl_);
+  }
+
+  struct PathAndFileSystem {
+    std::string path;
+    std::shared_ptr<fs::FileSystem> filesystem;
+  };
+
+  util::variant<PathAndFileSystem, std::shared_ptr<Buffer>, CustomOpen> impl_;
+  Compression::type compression_ = Compression::UNCOMPRESSED;
+};
+
+/// \brief The path and filesystem where an actual file is located or a buffer which can
+/// be written to like a file
+class ARROW_DS_EXPORT WritableFileSource {
```

Comment:
Instead of having to copy the class with the exact same code, can we stick with `OpenWritable()`?